pkg/controller/windowsmachine/metrics/metrics.go

```diff
@@ -0,0 +1,134 @@
+)
+
+// prometheusConfig holds the information required to configure prometheus, so that it can scrape metrics from the
+// given endpoint address
+type prometheusConfig struct {
+	// k8sclientset holds the information related to kubernetes clientset
+	k8sclientset *kubernetes.Clientset
+	// hostname defines name of the target node.
+	hostname string
+	// endpointAddress holds the information about the endpoint address in the endpoints object
+	endpointAddress *v1.EndpointAddress
+}
+
+// NewPrometheusConfig creates a new instance for prometheusConfig  to be used by the caller.
+func NewPrometheusConfig(clientset *kubernetes.Clientset, ipAddress, hostname string) *prometheusConfig {
+	endpointAddress := &v1.EndpointAddress{
+		IP:        ipAddress,
+		TargetRef: &v1.ObjectReference{Kind: "Node", Name: hostname},
+	}
+	return &prometheusConfig{
+		k8sclientset:    clientset,
+		hostname:        hostname,
+		endpointAddress: endpointAddress}
+
+}
+
+// Configure patches the endpoint object to reflect the current list Windows nodes. When addIPAddress is
+// true it adds the endpoint address for the current node, else it removes that endpoint address
+func (pc *prometheusConfig) Configure(addIPAddress bool) error {
+
+	// Retrieve endpoints object
+	endpoints, err := pc.k8sclientset.CoreV1().Endpoints(windowsMetricsEndpointsNamespace).Get(context.TODO(),
+		windowsMetricsEndpoints, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "error getting Endpoints object")
+	}
+
+	// Add Metrics Port if not present.
+	if len(endpoints.Subsets[0].Ports) == 0 || endpoints.Subsets[0].Ports[0].Port != 9182 {
+		patchData := fmt.Sprintf(
+			"[{\"op\": \"add\", \"path\": \"/subsets/0/ports/-\",\"value\":{\"name\": \"metrics\","+
+				"\"port\":%s,\"protocol\": %s}}]", string(metricsPort), v1.ProtocolTCP)
+
+		_, err := pc.k8sclientset.CoreV1().Endpoints(windowsMetricsEndpointsNamespace).Patch(context.TODO(),
+			windowsMetricsEndpoints, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if addIPAddress {
+		return pc.addEndpointAddress()
+	}
+	return pc.removeEndpointAddress()
+}
+
+// isNodeInEndpoints checks if the given node is present in the list of addresses in the endpoints object
+func (pc *prometheusConfig) isNodeInEndpoints() (string, error) {
+	endpoints, err := pc.k8sclientset.CoreV1().Endpoints(windowsMetricsEndpointsNamespace).Get(context.TODO(),
+		windowsMetricsEndpoints, metav1.GetOptions{})
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting endpoints object %v", windowsMetricsEndpoints)
+	}
+
+	foundNode := false
+	nodePosition := -1
+	for index, endpointAddress := range endpoints.Subsets[0].Addresses {
+		if pc.hostname == endpointAddress.TargetRef.Name {
+			foundNode = true
+			nodePosition = index
+			break
+		}
+	}
+	if !foundNode {
+		return "", errors.New(fmt.Sprintf("target node %s not found in Endpoints object %s",
```

Comment:
if this contains a format string, errors.wrap can be used.

Reply:
Kept it as is since I wanted error object

Reply:
`fmt.Errorf` is what we have been doing similar situations.